[RFC] Adding RTC API

jdascenzio opened the pull request:
Hi,
I'm working on a RTC driver for stm32. Currently in Zephyr, RTC is used as a counter. In my project, I would like to use it at a real clock time to timestamp my data. I also integrated this driver in  clock_[get/set]time. I'm open to all suggestions to improve it. In the future, I would like to integrate also alarm

tbursztyka commented on include/drivers/rtc.h:
```diff
@@ -0,0 +1,63 @@
+#include <time.h>
+#include <device.h>
+
+typedef int (*rtc_api_gettime)(struct device *dev, struct timespec *tp);
```
we usually add '_t' in the end of typedef-ed names

jdascenzio replied:
I fix it!

tbursztyka commented on lib/posix/clock.c:
```diff
@@ -34,6 +38,12 @@ int clock_gettime(clockid_t clock_id, struct timespec *ts)
 
 	case CLOCK_REALTIME:
+#ifdef CONFIG_RTC
+		dev_rtc = device_get_binding(DT_RTC_0_NAME);
```
it's going to be heavy, since it will lookup for the device each and every time clock_gettime() will be called.
Store dev_rtc statically in this file, and if NULL, then run device_get_binding()

jdascenzio replied:
You're right I fix it!

jdascenzio replied:
When I store dev_rtc statically, I have some troubles when clock_gettime is called in userspace. I try to find a solution

tbursztyka replied:
It's because you did not provide the drivers/rtc/rtc_handlers.c which should implement the userland "gateway" functions. See drivers/spi/spi_handlers.c for instance

jdascenzio replied:
I did it, but when clock_gettime is called, I have this error when my static dev_rtc is read
[00:00:00.005,000] <err> os: ***** MPU FAULT *****
[00:00:00.005,000] <err> os:   Data Access Violation
[00:00:00.005,000] <err> os:   MMFAR Address: 0x20000610
[00:00:00.005,000] <err> os: r0/a1:  0x00000001  r1/a2:  0x20001fc8  r2/a3:  0x00000000
[00:00:00.005,000] <err> os: r3/a4:  0x20000610 r12/ip:  0x08000b7b r14/lr:  0x08000bd7
[00:00:00.005,000] <err> os:  xpsr:  0x61000000
[00:00:00.005,000] <err> os: Faulting instruction address (r15/pc): 0x08008168
[00:00:00.005,000] <err> os: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
[00:00:00.005,000] <err> os: Current thread: 0x20000704 (unknown)
[00:00:00.078,000] <err> os: Halting system

tbursztyka replied:
I missed the addition, in which commit is it?

jdascenzio replied:
I push it!

jdascenzio replied:
I resolved this by turn Turn clock_[get/set]time into system calls

tbursztyka commented on lib/posix/clock.c:
```diff
@@ -74,6 +84,13 @@ int clock_settime(clockid_t clock_id, const struct timespec *tp)
+		return rtc_settime(dev_rtc, tp);
+	}
+#endif
+
```
not an issue in your commit, but line 118 seems bogus. The ARG_UNUSED should be on tz, not tv.

jdascenzio replied:
I made a pull request for that: https://github.com/zephyrproject-rtos/zephyr/pull/23550

nordic-krch commented on lib/posix/clock.c:
```diff
@@ -68,12 +91,22 @@ int clock_gettime(clockid_t clock_id, struct timespec *ts)
 		return -1;
 	}
 
+#ifdef CONFIG_RTC
```
you could consider using `IS_ENABLED` macro instead of ifdefs.

jdascenzio replied:
With ifdef, the code unneeded is removed..

nordic-krch replied:
same with IS_ENABLED() but then compiler removes it and not preprocessor. I has advantage that code is compiled even though there is no sample/test which enables given feature - no dead code.

jdascenzio replied:
Ok, it's a good reason!

nordic-krch commented on lib/posix/clock.c:
```diff
@@ -26,6 +40,9 @@ int clock_gettime(clockid_t clock_id, struct timespec *ts)
+#endif
 
 	switch (clock_id) {
 	case CLOCK_MONOTONIC:
```
so RTC is used only for REALTIME. Could that be an issue in some cases? It means that those clock ids are using different source clock thus accuracy. I'm not sure what are the typical uses cases but can imagine that someone reads both at one point of time and next time in an hour. Monotonic will show different time elapsed than realtime.

jdascenzio replied:
In my system I need to timestamp some data precisely.  My system could run during a long time without time synchronization. The RTC have a lower clock drift than system clock.

nordic-krch replied:
exactly, thus i wonder if it is possible to provide monotonic based on RTC device. Maybe logging uptime offset when settime is called?

jdascenzio replied:
There is 2 problems to use RTC as monotonic:
- According to Posix: "the MONOTONIC clock is not affected by discontinuous jumps in the system time". (A workaround will be to save the offset on settime and use it to correct monotonic clock)
- The resolution for shorter time is better with system clock

nordic-krch commented on include/drivers/rtc.h:
```diff
@@ -0,0 +1,65 @@
+};
+
+/**
+ * @brief Function to get RTC time
```
dots at the end of sentences. In parameters, return value too.

nordic-krch commented on lib/posix/clock.c:
```diff
@@ -34,6 +51,12 @@ int clock_gettime(clockid_t clock_id, struct timespec *ts)
 	case CLOCK_REALTIME:
+#ifdef CONFIG_RTC
+		dev_rtc = clock_getrtc();
+		if (dev_rtc) {
```
so if something goes wrong with rtc device we silently jump to use uptime. That should somehow be reported.

jdascenzio replied:
I rewrote this

nordic-krch commented on lib/posix/clock.c:
```diff
@@ -17,6 +18,19 @@
+	static struct device *rtc_dev;
+
+	if (rtc_dev == NULL) {
+		rtc_dev = device_get_binding(DT_RTC_0_NAME);
```
add at least LOG_ERR() when rtc_dev is null

jdascenzio replied:
I rewrote this

nordic-krch commented on lib/posix/clock.c:
```diff
@@ -17,6 +18,19 @@
+#ifdef CONFIG_RTC
+static struct device *clock_getrtc(void)
+{
+	static int isfirst = 1;
```
variable not used. It could be used to fetch device only once. Now if get binding returns null then it will always be called only to use uptime as clock source.

jdascenzio replied:
I removed it

tbursztyka commented on include/drivers/rtc.h:
```diff
@@ -0,0 +1,65 @@
+ *
+ * @return 0 on success, negative on error
+ */
+__syscall int rtc_gettime(struct device *dev, struct timespec *tp);
```
I forgot to mention: could you provide the drivers/rtc/rtc_handlers.c in a separate patch for userland support?

jdascenzio replied:
I did it!

tbursztyka commented on drivers/rtc/rtc_stm32.c:
```diff
@@ -0,0 +1,170 @@
+static int rtc_stm32_init(struct device *dev)
+{
+	int err = -EIO;
+	(void) dev;
```
if you are not using the dev parameter in this function, have ARG_UNUSED(dev) after variable declaration instead

jdascenzio replied:
I remove "(void) dev": dev is used after

jdascenzio commented on drivers/rtc/rtc_stm32.c:
```diff
@@ -0,0 +1,170 @@
+/*
```
I'd like to add alarm in my RTC driver. After, I think that the best would be that counter use RTC (maybe more work is need in this way but for me that make more sense). At the opposite, if RTC use counter, I have several problems:
- Counter init RTC to 0
- The counter have an accuracy only to second

erwango replied:
And also, RTC IP doesn't hold a counter register, so you'd need to convert time registers to counter value and then convert back counter to time... Highly inefficient.
Agree to keep them separated for now.

b0661 commented on drivers/rtc/rtc_stm32.c:
```diff
@@ -0,0 +1,170 @@
+	LL_RTC_TimeTypeDef timeDef;
+	LL_RTC_DateTypeDef dateDef;
+
+	LL_PWR_EnableBkUpAccess();
```
The path between LL_PWR_EnableBkUpAccess and LL_PWR_DisableBkUpAccess is a critical section. IMHO it should be protected by a e.g. semaphore preventing threads to access at the same time.

jdascenzio replied:
Now, I lock the scheduler to protect BkUpAccess and ensure that the new time is write quickly

jdascenzio committed:
drivers: rtc: add generic driver API header for rtc devices

API functions:
- rtc_gettime
- rtc_settime

Signed-off-by: Julien D'Ascenzio <julien.dascenzio@paratronic.fr>

jdascenzio committed:
drivers: rtc: stm32: add RTC driver

Supports STM32 RTC

Signed-off-by: Julien D'Ascenzio <julien.dascenzio@paratronic.fr>

jdascenzio committed:
lib: posix: clock: use RTC device

If an RTC device exit, il would be used for CLOCK_REALTIME

Signed-off-by: Julien D'Ascenzio <julien.dascenzio@paratronic.fr>

gmarull commented on soc/arm/st_stm32/stm32l4/soc.h:
```diff
@@ -71,7 +71,7 @@
 #endif
 
-#if defined(CONFIG_COUNTER_RTC_STM32)
+#if defined(CONFIG_COUNTER_RTC_STM32) || defined(CONFIG_RTC_STM32)
```
This should likely be enabled on more SoC if driver is generic enough

erwango replied:
I'm fine to limit on L4 to start with. There is sufficient work to introduce the api and the driver.
Enabling (and validating) on more SoC could be done later on.

jdascenzio committed:
drivers: rtc: add system call handlers

Add system call handlers

Signed-off-by: Julien D'Ascenzio <julien.dascenzio@paratronic.fr>

jdascenzio committed:
CODEOWNERS: add jdascenzio to /drivers/rtc/

Add jdascenzio to be automatically added for reviews
for changes in /drivers/rtc/

Signed-off-by: Julien D'Ascenzio <julien.dascenzio@paratronic.fr>

tbursztyka commented on drivers/rtc/CMakeLists.txt:
```diff
@@ -3,3 +3,5 @@
 
 zephyr_library_sources_ifdef(CONFIG_RTC_STM32           rtc_stm32.c)
+
+zephyr_sources_ifdef(CONFIG_USERSPACE   rtc_handlers.c)
```
zephyr_library_sources_ifdef() here

tbursztyka commented on drivers/rtc/rtc_handlers.c:
```diff
@@ -0,0 +1,22 @@
+#include <syscall_handler.h>
+#include <drivers/rtc.h>
+
+static inline int z_vrfy_rtc_gettime(struct device *dev, struct timespec *tp)
```
I think you'll need to copy tp

jdascenzio replied:
Why I need to copy? However, I could had:
Z_OOPS(Z_SYSCALL_MEMORY_WRITE(tp, sizeof(*tp)));

tbursztyka commented on drivers/rtc/rtc_handlers.c:
```diff
@@ -0,0 +1,22 @@
+#include <syscalls/rtc_gettime_mrsh.c>
+
+static inline int z_vrfy_rtc_settime(struct device *dev, const struct timespec *tp)
+{
```
same here

jdascenzio replied:
Here: Z_OOPS(Z_SYSCALL_MEMORY_READ(tp, sizeof(*tp)));

jdascenzio commented on drivers/rtc/rtc_stm32.c:
```diff
@@ -0,0 +1,170 @@
+
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(rtc_stm32, LOG_LEVEL_DBG);
```
ok

b0661 commented on drivers/rtc/rtc_stm32.c:
```diff
@@ -0,0 +1,182 @@
+	k_sched_unlock();
+
+	if (serr) {
+		LOG_ERR("failed to set date");
```
Should be LOG_ERR(serr)

erwango commented on drivers/rtc/rtc_stm32.c:
```diff
@@ -0,0 +1,182 @@
+
+#if defined(CONFIG_RTC_STM32_CLOCK_LSI)
+
+#if defined(CONFIG_SOC_SERIES_STM32WBX)
```
Please remove any code specific to series that are not L4.

jdascenzio replied:
It's done!

erwango commented on drivers/rtc/rtc_stm32.c:
```diff
@@ -0,0 +1,182 @@
+		goto out;
+	}
+
+	if (LL_RTC_BAK_GetRegister(RTC, LL_RTC_BKP_DR0) != 0x32F2) {
```
Please avoid magic numbers

jdascenzio replied:
I copied this from STM32 example (RTC_Calendar). I change this with the flag INITF (Initialization flag) of RTC_ISR register. I think it's more appropriate.

erwango commented on drivers/rtc/rtc_stm32.c:
```diff
@@ -0,0 +1,182 @@
+/* prescaler values for LSI @ 32 KHz */
+#define RTC_PREDIV_ASYNC 31
+#define RTC_PREDIV_SYNC  999
+#else /* CONFIG_COUNTER_RTC_STM32_CLOCK_LSE */
```
CONFIG_RTC_STM32_CLOCK_LSE

jdascenzio replied:
I fixed it!

erwango reviewed (changes requested):
Thanks for this addition!

Since you're adding a new API, can you add a matching basic api test under `tests/drivers/...`?

b0661 commented on drivers/rtc/rtc_stm32.c:
```diff
@@ -0,0 +1,180 @@
+	LOG_DBG("RTC initialised");
+	err = 0;
+out:
+	LL_PWR_DisableBkUpAccess();
```
Can´t LOG_DBG be postponed after LL_PWR_DisableBkUpAccess()?

jdascenzio replied:
I did it

jdascenzio committed:
boards: arm: stm32l4: add rtc feature

add rtc feature to various stm32l4 boards

Signed-off-by: Julien D'Ascenzio <julien.dascenzio@paratronic.fr>

jdascenzio committed:
tests: drivers: add test for rtc

Add tests for rtc driver
Test on nucleo-l452re

Signed-off-by: Julien D'Ascenzio <julien.dascenzio@paratronic.fr>

jdascenzio commented:
I made this change:
 - Add CONFIG_RTC_LOG_LEVEL
 - add rtc features on some boards
 - add driver rtc api test
 - rtc_stm32_gettime: fix bug on subSeconds calculation

I created a pull request (#23630) for commit d9fb1f5f19352e8e96db4786dd2329f520184114

jdascenzio commented:
Now, I used snake case for rtc driver api .
I removed commit about clock_[get/set]time into system calls

erwango commented on tests/drivers/rtc/api/src/main.c:
```diff
@@ -0,0 +1,58 @@
+	/* Set a particular time: 2020-03-23 12:22:40 */
+	ts0.tv_sec = 1584966160;
+	ts0.tv_nsec = 0;
+	ret = rtc_settime(dev, &ts0);
```
`rtc_set_time`

jdascenzio replied:
Thanks, I fix it!

erwango commented on tests/drivers/rtc/api/src/main.c:
```diff
@@ -0,0 +1,58 @@
+
+	k_sleep(waitTime_ms);
+
+	ret = rtc_gettime(dev, &ts);
```
`rtc_get_time`

jdascenzio replied:
Thanks, I fix it!

jdascenzio commented:
Fix some indents

pabigot commented on include/drivers/rtc.h:
```diff
@@ -0,0 +1,67 @@
+
+/**
+ * @brief Function to get RTC time
+ *
```
IMO it's worth giving a little bit more information here about the time.  Specifically I would state that it is applications-specific whether the time maintained is UTC, or offset by a particular zone, or subject to daylight saving adjustments.  Also probably necessary to note that whether time is maintained across sleep or deep-sleep periods.

That's something that should be discussed at the API level.

jdascenzio replied:
The time represents seconds and nanoseconds since the Epoch. I don't think that localtime must be managed by the RTC driver

pabigot replied:
Which epoch?
* 1970-01-01T00:00:00Z (POSIX)
* 1980-01-05T00:00:00Z (GPS)
* ?? (Bluetooth mesh)

Does the timer advance in UTC seconds? TAI seconds?  They're not the same.

jdascenzio replied:
For me, epoch is UNIX timestamp like clock_gettime with CLOCK_REALTIME. So, the timer advance in UTC seconds.

pabigot replied:
Advancing in UTC seconds introduces complexity if you have a local RTC that increments based on experienced (TAI) seconds, because you have to adjust based on leap seconds which are announced every 6 months.  The timer is technically monotonic, but does not have a constant slope.  Calculating the duration in experienced seconds between two UTC times is non-trivial.  Supporting it would require functions to detect the number of leap seconds in effect at the time, which depends on a table.  Local time zone is similarly complex, but can derive from UTC given a different table.

For the underlying counter TAI seconds after the POSIX epoch may be the most flexible fixed specification, but the most general is likely to be a description of the time scale given a civil time representation of the epoch (in UTC) and whether the stored time already accounts for subsequent leap seconds or time zone offsets.

pabigot commented on include/drivers/rtc.h:
```diff
@@ -0,0 +1,67 @@
+ * @param[in] dev RTC device
+ * @param[in] tp pointer to new rtc value
+ *
+ * @return 0 on success, negative on error
```
This is probably the minimal necessary.  For precision timing you want a way to correlate the preserved time with the local system clock in ticks.  https://github.com/zephyrproject-rtos/zephyr/pull/17631 has some API that does that.

jdascenzio replied:
This synchronization is interesting. In my opinion, it must be done regularly in the driver (according to the drift of system clock) to permit the [get / set] _time functions to have more precision. I don't think that exposed some synchronization API is necessary.

pabigot replied:
I don't think adjustment belongs in the RTC driver.  The local RTC itself may be corrected based on an external time source like PTP, NTP, GPS, or mesh time service, and many will have device-specific mechanisms to tune the rate.

The ability to align the system tick clock with each time source independently (for both sets and gets) allows you to transfer times from one source to another with minimal error and without burdening each time-related component with extra code trying to compensate for anticipated error.

The tick drift may not be constant (temperature effects) so I expect it to be maintained separately with its own infrastructure, retrieving synchronized time values from different sources and adjusting the tick-to-time offset parameters dynamically.  We're not at that level of maturity in Zephyr yet.

jdascenzio replied:
Thank you for this exhaustive response. It is true that in my current case, I wish to have only one precise time reference to be able to timestamp my data. My approach to the subject is therefore surely a little simplistic. I don't have enough experience in clocks managing and synchronization mechanisms to know which structure would be suitable for this.

pabigot commented on lib/posix/clock.c:
```diff
@@ -34,6 +45,9 @@ int clock_gettime(clockid_t clock_id, struct timespec *ts)
 		break;
 
 	case CLOCK_REALTIME:
+		if (rtc_dev) {
```
I suspect this is OK, but somebody might want to make it optional, especially if the RTC maintains whole-second resolution.

jdascenzio replied:
Indeed, I can use a different option for that like CONFIG_USE_RTC_FOR_CLOCK_REALTIME

pabigot reviewed (commented):
Just an initial review for now.  If this gets traction I'd like to see it evolve in conjunction with #17631.

carlescufi commented:
API Meeting:

- A combination of the counter API + translation functions (from ticks to civil time) will already cover this usecase
- We need more implementations of RTCs using the counter API before we can actually proceed to extract the common parts into its own API
- We need more input from users and contributors interested in RTC in order to make the right decision

Proposal: continue to use this PR to discuss what an RTC API should look like alongside the API meeting. @tbursztyka @pabigot @nordic-krch @tbursztyka @jdascenzio: let's continue the discussion here and please join the API meeting if possible to find out what the API should look like.

FRASTM commented on drivers/rtc/rtc_stm32.c:
```diff
@@ -0,0 +1,181 @@
+#define RTC_PREDIV_ASYNC 31
+#define RTC_PREDIV_SYNC  999
+#else /* CONFIG_RTC_STM32_CLOCK_LSE */
+/* prescaler values for LSI @ 32768 Hz */
```
prescaler values for LS**E**

FRASTM commented on drivers/rtc/rtc_stm32.c:
```diff
@@ -0,0 +1,181 @@
+#include <logging/log.h>
+LOG_MODULE_REGISTER(rtc);
+
+#if defined(CONFIG_RTC_STM32_CLOCK_LSI)
```
Please give the targetted rtc clock 
```
(The asynchronous prescaler division factor is set to 128, and the synchronous division
factor to 256, to obtain an internal clock frequency of 1 Hz (ck_spre) with an LSE frequency
of 32.768 kHz.)
```

b0661 commented on drivers/rtc/rtc_stm32.c:
```diff
@@ -0,0 +1,181 @@
+
+	err = 0;
+out:
+	LL_PWR_DisableBkUpAccess();
```
Looks a little bit strange to enable something just to be able to disable it. Wouldn´ t an else path be more appropriate?

b0661 commented on drivers/rtc/rtc_stm32.c:
```diff
@@ -0,0 +1,181 @@
+out:
+	LL_PWR_DisableBkUpAccess();
+	if (err == 0)
+		LOG_DBG("RTC initialised correctly");
```
Coding convention for Zephyr requests braces.

henrikbrixandersen commented:
I have an RTC driver pending too. I was planning to suggest an extension of the counter API to cover the civic time part of the RTC. I will try to finalise my take on this for the next API meeting.

jdascenzio commented:
> I have an RTC driver pending too. I was planning to suggest an extension of the counter API to cover the civic time part of the RTC. I will try to finalise my take on this for the next API meeting.

Have you finished your patch? I would be interested in your extension solution. For me, the problems I have with the counter API is:
- The type of counters is int which is a problem for 32bit platforms with 2038 bug.
- There is no API to set the current value

I wrote also an RTC driver for the microcrystal rv3029 component. I'm waiting to see which RTC management solution is chosen to offer it to the community

jukkar commented on drivers/rtc/CMakeLists.txt:
```diff
@@ -3,3 +3,5 @@
 zephyr_library()
```
In commit `drivers: rtc: add system call handlers`, I can understand that it is easy to just copy the commit subject to the body but we should try to avoid this. It is really annoying later if one looks the commits and tries to understand what the commit is about and both the subject and body are same (not really applicable for this commit as it is easy to understand what the commit does but as a general note). For example here the commit body could say `Add system call handlers so that the RTC get/set functions can be called from userspace`. So the commit body should tell why we are doing the change, it is usually not necessary to document what the commit does as that can be usually seen from the code.

sslupsky commented:
Regarding the rtc api, I would appreciate if we could maintain the clock through a mcu reboot.  Please see #24625

sslupsky commented:
Please also add this api should consider those situations where the rtc is used to generate the system tick.

It may also be helpful to consider rtc timers.  Some soc rtc's can generate one or more interrupts from the rtc either directly or using the event system.  A setup and callback api (similar to K_TIMER?) would be useful.

erwango commented:
> his api should consider those situations wh

@sslupsky can you check that current RTC based counter driver is not already answering this particular need?

sslupsky commented:
@erwango At the moment, there is no rtc counter driver for the sam0 in drivers/counter.  The system timer for the sam0 uses the soc rtc though (drivers/timer/sam0_rtc_timer.c).  I am recommending we consider this dual purpose in the scope of the rtc api since the rtc could be used for both purposes.

sslupsky commented:
The TI cc13x2 and cc26x2 sac's use an rtc for the system timer as well (drivers/timer/cc13x2_cc26x2_rtc_timer.c).

sslupsky commented:
I did a review of #17631 and an item comes to mind related to the rtc api discussion here.  We should probably define what value the rtc counter stores?  Is it UTC?  If so we need to have an api that informs about location so that local standard time / civil time can be derived.  We probably need hooks in there to coordinate with GNSS as well.  Which brings up another zephyr api requirement highly related to time which is GNSS location.

Another thought is to provide for a rtc temperature compensation api.

sslupsky commented:
@pabigot Yes, I think you are correct about the set of libraries.  It would be great if these libraries could detect changes and adapt and report time accordingly.  For example, geo location informing the time library would be helpful.

jdascenzio commented:
At this time, I have some issues with the counter API for these reasons:
- The counter is reset to 0 during initialization (at least for the STM32). I need to keep the datetime after a power cut or after waking up from a deep sleep mode
- It isn’t possible to set a value
- The counter is only 32bit: The precision is only at the second and there is 2038 bug for unix epoch time
These are the main reasons that led me to propose this API.

pabigot commented:
> At this time, I have some issues with the counter API for these reasons:
> 
> * The counter is reset to 0 during initialization (at least for the STM32). I need to keep the datetime after a power cut or after waking up from a deep sleep mode

Agreed.  There must be some way to specify that the counter should not be reset (and, if running, should not be stopped) during driver initialization.

It may be sufficient to just document that as an expectation, and correct this behavior in STM32.  Gratuitously resetting a running counter on initialization seems like bad idea in general and for low-power devices specifically, regardless of whether what it's counting is the passage of time.

> * It isn’t possible to set a value

That can be fixed by adding a `counter_set_value()` function that is implemented where the hardware allows it, understanding that doing so probably stops and restarts the counter.

It could also be supported without mutating the running counter value by instead adding API to store a (signed) offset that's added to the current counter value whenever it's read (through a new `counter_get_adjusted_value()` API).

> * The counter is only 32bit: The precision is only at the second and there is 2038 bug for unix epoch time

Agreed, but again we could add `counter_get_value_64()` for drivers that support rollover into a larger representation; and/or integrate that with an API that manages offsets.

The original Zephyr RTC and COUNTER APIs were discarded when the new COUNTER API was added in part because the differences between RTC and COUNTER weren't considered important.  I think the RTC use case could have had a larger impact on the original design of COUNTER, and it may well turn out that COUNTER isn't the right core API, but for now that underlying similarity is pretty strong.

sslupsky commented:
> It may be sufficient to just document that as an expectation, and correct this behavior in STM32. Gratuitously resetting a running counter on initialization seems like bad idea in general and for low-power devices specifically, regardless of whether what it's counting is the passage of time.

Agreed ... but ... I considered what you suggested here but bumped into the definition of the system timer.  It appears the zephyr system timer api defines that time starts at zero at boot?  Can that be changed?  If this definition is set in stone we probably need to consider multiple times, the zephyr system time and the real time(s) and have api's to synchronize them and maintain offsets at boot, time source (ntp), geo location (timezone, gps, wifi, etc), another counter (ds3231, etc), etc.

pabigot commented:
> Agreed ... but ... I considered what you suggested here but bumped into the definition of the system timer. It appears the zephyr system timer api defines that time starts at zero at boot? Can that be changed?

IMO no, the system time should always represent time-since-startup (with some wiggle for deep-sleep periods).  Real time is a different clock.

pabigot commented:
API 2020-07-07: Moved from triage to To-Do pending submission of draft proposals by @pabigot, @henrikbrixandersen, or others.

llisauskas commented:
@jdascenzio is this PR going to recieve any updates? I'm planning to start work on something similiar, since my system require something similiar, but it does not make sense if this is going forward.

jdascenzio commented:
> @jdascenzio is this PR going to recieve any updates? I'm planning to start work on something similiar, since my system requires this, but it does not make sense if this is going forward. I can contribute if it get's approved.

For now, I have stopped to work on it. I used this driver which suits me for the moment. Maybe @pabigot or @henrikbrixandersen will make a draft to add some of these functionalities in driver counter....

pabigot commented:
> For now, I have stopped to work on it. I used this driver which suits me for the moment. Maybe @pabigot or @henrikbrixandersen will make a draft to add some of these functionalities in driver counter....

The architecture design I still support is that counters continue to provide an integer interface, and things like `timeutil_timegm64` and `gmtime` are used to translate between civil and integer time representations to allow a counter to support civil time maintenance and alarms.  There are also supporting capabilities necessary to:
1. register civil time against the local clock (without that being restricted to `CONFIG_POSIX=y`);
2. accommodate leap seconds and local time offset changes for people who require them; and
3. compensate for local clock skew.

I have draft support for the pieces of that that are important to me (1 and 3), but they didn't make the 2.4.0 deadline.  I expect to get them posted around the time 2.5.0 development opens up in a few weeks.

cederom commented:
Hello world :-) I also create set time by manual time shift on set add.. but this is not preserved across reboots. I have created a feature request for RTC framework here: https://github.com/zephyrproject-rtos/zephyr/issues/35333 :-)
